keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/PartialDerivative.java

```diff
@@ -0,0 +1,170 @@
+package io.improbable.keanu.vertices.dbl.nonprobabilistic.diff;
+
+import io.improbable.keanu.tensor.TensorShape;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+
+import java.util.Arrays;
+
+public class PartialDerivative {
```

Comment:
I know we might add a bit more state to this in future to help with tracking dimensions etc - but for now it could be argued that having this as an explicit object is just costing us in terms of performance (due to the extra level of indirection around the Tensor object).  I think it's fine for now, but something to consider in future that given every method call has to go via this object and then to the tensor object we're hopping around memory quite a bit.

Reply:
It's not ideal but it provides some autodiff context that DoubleTensor wouldn't give us. If we want to track the shape wrt/of then we'll need a class to hold it and it's a nice way to represent a null partial derivative. The null partial could probably be replaced with some clever refactoring but the wrt/of shape is a autodiff domain concept that doesn't map to DoubleTensor. If we move towards graphifying the autodiff then this indirection will be compiled out anyways.

Reply:
indeed - as I say, not a prob for now.